### PR TITLE
genesys: usbhub: support hid isp

### PR DESCRIPTION
**Pull request by adamgene**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation
---

To send a vendor command through child HID. We use HID set/get feature report, and implement multiple read/write.

### CODE REVIEW

File: plugins/genesys/fu-genesys-hubhid-device.c [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+		g_warning("read zero-length hid report");
+		return TRUE;
+	}
+
+	/* receive report */
+	chunks = fu_chunk_array_mutable_new(data, setup->length, 0, 0x0, self->buf_report->len);
+	if (progress != NULL) {
+		fu_progress_set_id(progress, G_STRLOC);
+		fu_progress_set_steps(progress, chunks->len);
+	}
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-hubhid-device.c:68]: I think we should always use a FuProgress

File: plugins/genesys/fu-genesys-hubhid-device.c [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+#include "fu-genesys-hubhid-device.h"
+
+#define GENESYS_HUBHID_REPORT_ID	  0
+#define GENESYS_HUBHID_REPORT_BYTE_LENGTH 0x40
+#define GENESYS_HUBHID_REPORT_TIMEOUT	  100 /* ms */
+#define GENESYS_HUBHID_REPORT_FLAGS	  FU_HID_DEVICE_FLAG_ALLOW_TRUNC | FU_HID_DEVICE_FLAG_IS_FEATURE
+
+struct _FuGenesysHubhidDevice {
+	FuHidDevice parent_instance;
+	GByteArray *buf_report;
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-hubhid-device.c:21]: it seems a bit weird to put this as device instance data... it seems like the kind of thing that should be returned from fu_genesys_hubhid_device_command_read()

File: plugins/genesys/fu-genesys-plugin.c [hunk 3]
```diff
@@ -35,13 +36,52 @@ fu_genesys_plugin_constructed(GObject *obj)
...
+			FuDevice *dev = g_ptr_array_index(devices, i);
+
+			if (!FU_IS_GENESYS_USBHUB_DEVICE(dev))
+				continue;
+
+			if (fu_usb_device_get_dev(FU_USB_DEVICE(dev)) == gusb_parent) {
+				parent = dev;
+				break;
+			}
+		}
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-plugin.c:68]: i think split this out to something like `FuDevice *fu_genesys_plugin_get_device_by_gusb_device(self, GUsbDevice *gusb_device)`

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 4]
```diff
@@ -137,10 +140,82 @@ struct _FuGenesysUsbhubDevice {
...
 };
 
 G_DEFINE_TYPE(FuGenesysUsbhubDevice, fu_genesys_usbhub_device, FU_TYPE_USB_DEVICE)
 
+void
+fu_genesys_usbhub_device_set_hid_channel(FuDevice *device, FuDevice *channel)
+{
+	FuGenesysUsbhubDevice *self = FU_GENESYS_USBHUB_DEVICE(device);
+	if (self == NULL)
+		return;
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:155]: why would self be NULL? I think this would be more a g_return_if_fail()

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 5]
```diff
@@ -137,10 +140,82 @@ struct _FuGenesysUsbhubDevice {
...
+static gboolean
+fu_genesys_usbhub_device_send_control_transfer(FuGenesysUsbhubDevice *self,
+					       FuGenesysUsbSetup *setup,
+					       guint8 *data,
+					       FuProgress *progress,
+					       GError **error)
+{
+	if (self->hid_channel != NULL) {
+		guint8 dir = setup->req_type.dir;
+		setup->req_type.dir = !dir; /* revert g_usb in/out dir to usb spec */
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:179]: ??

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 6]
```diff
@@ -137,10 +140,82 @@ struct _FuGenesysUsbhubDevice {
...
+				    self->hid_channel,
+				    setup,
+				    data,
+				    error);
+			} else {
+				return fu_genesys_hubhid_device_command_write(self->hid_channel,
+									      setup,
+									      data,
+									      progress,
+									      error);
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:199]: i think this should be handled in fu_genesys_hubhid_device_command_write() when you pass in the GByteArray as a parameter.

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 7]
```diff
@@ -137,10 +140,82 @@ struct _FuGenesysUsbhubDevice {
...
+		}
+	} else {
+		return g_usb_device_control_transfer(fu_usb_device_get_dev(FU_USB_DEVICE(self)),
+						     setup->req_type.dir,
+						     setup->req_type.type,
+						     setup->req_type.recipient,
+						     setup->request,
+						     setup->value,  /* value */
+						     setup->index,  /* idx */
+						     data,	    /* data */
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:210]: no sure you need these three comments :)

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 8]
```diff
@@ -161,31 +236,34 @@ fu_genesys_usbhub_device_read_flash(FuGenesysUsbhubDevice *self,
...
-						   (fu_chunk_get_address(chk) & 0x0f0000) >>
-						       4,			       /* value */
-						   fu_chunk_get_address(chk) & 0xffff, /* idx */
-						   fu_chunk_get_data_out(chk),	       /* data */
-						   fu_chunk_get_data_sz(chk), /* data length */
-						   NULL,		      /* actual length */
-						   GENESYS_USBHUB_USB_TIMEOUT,
-						   NULL,
-						   error)) {
+		g_autofree FuGenesysUsbSetup *setup = g_malloc0(sizeof(FuGenesysUsbSetup));
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:252]: g_new0() here please

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 9]
```diff
@@ -161,31 +236,34 @@ fu_genesys_usbhub_device_read_flash(FuGenesysUsbhubDevice *self,
...
+
+		setup->req_type.dir = G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST;
+		setup->req_type.type = G_USB_DEVICE_REQUEST_TYPE_VENDOR;
+		setup->req_type.recipient = G_USB_DEVICE_RECIPIENT_DEVICE;
+		setup->request = self->vcs.req_read;
+		setup->value = (fu_chunk_get_page(chk) & 0x000f) << 12;
+		setup->index = fu_chunk_get_address(chk) & 0xffff;
+		setup->length = fu_chunk_get_data_sz(chk);
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:260]: does using the USB names add clarity here, or make it worse? I'd argue a custom struct with setup->address = fu_chunk_get_address(chk) would be easier to understand -- i.e. you store them in the struct what they *are* not what you're sending them *as*

File: plugins/genesys/fu-genesys-usbhub-device.c (hunk 5 above)
Comment (adamgene) [reply plugins/genesys/fu-genesys-usbhub-device.c:179]: The define does not match the common USB BM_REQUEST_TYPE https://github.com/fwupd/fwupd/blob/651b6b1207ae923e2368d28c5cd88e02a4ec501d/plugins/genesys/fu-genesys-common.h#L22-L30

`GUsbDeviceDirection` is the inverse of with [wdfusb](https://github.com/microsoft/Windows-Driver-Frameworks/blob/a94b8c30dad524352fab90872aefc83920b98e56/src/publicinc/wdf/kmdf/1.33/wdfusb.h#L50-L56), that cause `G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE` set to 0xC0(vendor read).

File: plugins/genesys/fu-genesys-hubhid-device.c (hunk 2 above)
Comment (adamgene) [reply plugins/genesys/fu-genesys-hubhid-device.c:21]: The `buf_report` is used to send/receive GL report, and the size should be queried by HID report descriptor(the report size must fixed as the descriptor declared). Currently I set it a hardcode value to go.

File: plugins/genesys/fu-genesys-usbhub-device.c (hunk 9 above)
Comment (adamgene) [reply plugins/genesys/fu-genesys-usbhub-device.c:260]: > does using the USB names add clarity here, or make it worse?

I think this function already named the purpose, and `fu_genesys_usbhub_device_send_control_transfer` too. Therefore, add a middle-function to do all the R/W again, maybe cause some confuse?

Comment (hughsie) [reply plugins/genesys/fu-genesys-usbhub-device.c:260]: Yup, maybe it's confusing.

File: plugins/genesys/fu-genesys-usbhub-device.c (hunk 5 above)
Comment (adamgene) [reply plugins/genesys/fu-genesys-usbhub-device.c:179]: `FuGenesysUsbRequestType` is using to align the USB spec, that is needed to fulfill format of GL HID feature report. 

The GL HID feature report format has these phases:
1. send request command
`setup`: Set feature report
`data`:  original USB vendor command that we send by g_usb, i.g. `40 81 0100 0000 0000`
2. read/write data
    - `read data`
        - `setup`: Get feature report
        - `data`    
    - `write data`
        - `setup`: Set feature report
        - `data`
3. send end to finish write
`setup`: Get feature report

File: plugins/genesys/fu-genesys-hubhid-device.c (hunk 1 above)
Comment (adamgene) [reply plugins/genesys/fu-genesys-hubhid-device.c:68]: I try to modify to give each command with a FuProgress, and I have a question. Does it need a FuProgress if read only one byte?

Comment (hughsie) [reply plugins/genesys/fu-genesys-hubhid-device.c:68]: depends on how long that read takes :) If it's less than 1ms then don't bother.

File: plugins/genesys/fu-genesys-hubhid-device.c (hunk 2 above)
Comment (hughsie) [reply plugins/genesys/fu-genesys-hubhid-device.c:21]: I think storing the size makes sense, just much less the buffer. It's acting as a cache.

File: plugins/genesys/fu-genesys-hubhid-device.c [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+{
+	const guint16 hid_token[] = {
+	    GUINT16_TO_LE('G'),
+	    GUINT16_TO_LE('L'),
+	    GUINT16_TO_LE('I'),
+	    GUINT16_TO_LE(' '),
+	    GUINT16_TO_LE('H'),
+	    GUINT16_TO_LE('I'),
+	    GUINT16_TO_LE('D'),
+	};
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-hubhid-device.c:248]: do you want g_utf8_to_utf16 here?

Comment (hughsie) [reply plugins/genesys/fu-genesys-hubhid-device.c:248]: Or there's fu_utf8_to_utf16_byte_array() too

File: plugins/genesys/fu-genesys-hubhid-device.c [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+
+static gboolean
+fu_genesys_hubhid_device_open(FuDevice *device, GError **error)
+{
+	/* FuHidDevice->open */
+	if (!FU_DEVICE_CLASS(fu_genesys_hubhid_device_parent_class)->open(device, error))
+		return FALSE;
+
+	return TRUE;
+}
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-hubhid-device.c:304]: do we need this?

File: plugins/genesys/fu-genesys-hubhid-device.c [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+
+static gboolean
+fu_genesys_hubhid_device_close(FuDevice *device, GError **error)
+{
+	/* FuHidDevice->close */
+	if (!FU_DEVICE_CLASS(fu_genesys_hubhid_device_parent_class)->close(device, error))
+		return FALSE;
+
+	return TRUE;
+}
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-hubhid-device.c:314]: same

File: plugins/genesys/fu-genesys-hubhid-device.c [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+		return FALSE;
+
+	/* pack report if it can */
+	pack_report = fu_genesys_hubhid_device_can_pack_report(self, setup->length);
+	if (pack_report) {
+		if (setup->length > 0 && !fu_memcpy_safe(buf_report->data,
+							 buf_report->len,
+							 sizeof(FuGenesysUsbSetup), /* dst */
+							 data,
+							 setup->length,
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-hubhid-device.c:155]: is data always of length  setup->length? I'd feel a lot more comfortable with a parameter of `gsize datasz` to this function.

File: plugins/genesys/fu-genesys-hubhid-device.c (hunk 11 above)
Comment (adamgene) [reply plugins/genesys/fu-genesys-hubhid-device.c:304]: The `hubhid` is a command channel for hub, and we keep operating to hub.
I have tried using proxy for hub first, but I misunderstand the proxy mechanism.
Therefore, I add `->open()`/`->close()` in `usbhub` and `hubhid` to open the hid device.
https://github.com/fwupd/fwupd/blob/dc5c837ece66c64ff009b5e7b840b85b61b413ac/plugins/genesys/fu-genesys-usbhub-device.c#L1397-L1431

File: plugins/genesys/fu-genesys-hubhid-device.c (hunk 13 above)
Comment (adamgene) [reply plugins/genesys/fu-genesys-hubhid-device.c:155]: > is data always of length setup->length?

data may larger than setup->length.

>  I'd feel a lot more comfortable with a parameter of gsize datasz to this function.

TL;DR:  I will check all vendor command caller to meet the goal of your wish.

Because this function is called from `fu_genesys_usbhub_device_ctrl_transfer`, and all vendor command calls `fu_genesys_usbhub_device_ctrl_transfer`. I think the vendor command caller has responsibility to maintain the `data` and `datasz`. 

I can add a comment for these two functions to indicate user has the responsibility of `data` and `setup->length`.

Comment (adamgene) [reply plugins/genesys/fu-genesys-hubhid-device.c:173]: I changed the `fu_genesys_usbhub_device_ctrl_transfer` interface to fulfill `g_usb_device_control_transfer`, and revert the caller usage. Can you check that is well?

File: plugins/genesys/fu-genesys-hubhid-device.c (hunk 1 above)
Comment (hughsie) [reply plugins/genesys/fu-genesys-hubhid-device.c:91]: Still valid, no?

File: plugins/genesys/fu-genesys-hubhid-device.c [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+	guint16 report_length;
+	guint16 max_report_pack_data_length;
+};
+
+G_DEFINE_TYPE(FuGenesysHubhidDevice, fu_genesys_hubhid_device, FU_TYPE_HID_DEVICE)
+
+static gboolean
+fu_genesys_hubhid_device_command_read(FuGenesysHubhidDevice *self,
+				      FuGenesysUsbSetup *setup,
+				      guint8 *data,
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-hubhid-device.c:50]: i think include `gsize datasz` and then `g_return_val_if_fail(datasz == setup->length, FALSE)`

File: plugins/genesys/fu-genesys-hubhid-device.c [hunk 15]
```diff
@@ -0,0 +1,400 @@
...
+	setup->index = 0;
+	setup->length = 0x40;
+
+	fu_byte_array_set_size(buf_data, setup->length, 0);
+
+	if (!fu_genesys_hubhid_device_command_read(self, setup, buf_data->data, NULL, error))
+		return FALSE;
+	if (!fu_memcmp_safe(buf_data->data,
+			    buf_data->len,
+			    2,
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-hubhid-device.c:314]: i think offsets we're normally specifying like 0x2

File: plugins/genesys/fu-genesys-hubhid-device.c (hunk 11 above)
Comment (hughsie) [reply plugins/genesys/fu-genesys-hubhid-device.c:353]: Right, but I think the code you've got above is a NOP -- i.e. if you don't override FuGenesysHubhidDevice->open then it'll fall back to FuHidDevice->open - i.e. you could delete the ->open function and it'd work the same as now.

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 16]
```diff
@@ -1234,6 +1342,10 @@ fu_genesys_usbhub_device_attach(FuDevice *device, FuProgress *progress, GError *
 	if (!fu_genesys_usbhub_device_reset(self, error))
 		return FALSE;
 
+	if (self->hid_channel != NULL) {
+		fu_device_add_flag(FU_DEVICE(self->hid_channel), FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	}
+
 	/* success */
 	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:1350]: shouldn't this be or? as in put the `fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);` in an `else`

File: plugins/genesys/fu-genesys-usbhub-device.c [hunk 17]
```diff
@@ -1470,6 +1660,7 @@ fu_genesys_usbhub_device_setup(FuDevice *device, GError **error)
 				    "VENDORSUP",
 				    NULL);
 	fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "PUBKEY", NULL);
+	fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "PROJECT", NULL);
```
Comment (hughsie) [inline plugins/genesys/fu-genesys-usbhub-device.c:1663]: I think this needs documenting in the plugin README.md

File: plugins/genesys/fu-genesys-usbhub-device.c (hunk 16 above)
Comment (adamgene) [reply plugins/genesys/fu-genesys-usbhub-device.c:1350]: The HID is the substitute of hub; after sent reset by hid, hub also reset.

File: plugins/genesys/fu-genesys-hubhid-device.c (hunk 1 above)
Comment (adamgene) [reply plugins/genesys/fu-genesys-hubhid-device.c:91]: No. I already checked all the R/W command, and add the `FuProgress` if it takes more than 1ms.

**Review by adamgene — COMMENTED** [review]
I have rebased and squashed the commits, please check that. THX.
